File: src/evaluation/PartInfoZero.m
% This is similar to PartInfo for level >= 3, but specialized for (i, j, k) containing zeros.
% In this case, many variables are not used. We do not register for them.
% A static method in PartInfo will decide which class to create, PartInfo or PartInfoZero, or even PartInfoLv2.

classdef PartInfoZero < matlab.mixin.Copyable
properties
  level
  power
  sum_col
  sum_half
  part_id
  shape
  zero_dim  % Which dimension is zero.
  nonzero_dim_1
  nonzero_dim_2

  % Optimizable vars.
  complete_split  % Output 1x3 cell array of CSD. We give individual IDs to the optimizable ones within them.
  complete_split_id  % 1x3 cell array of scalar row vectors of IDs in param_manager. -1 represents "always zero". -2 represents "always one". Different elements can share the same ID meaning that they are always equal.

  % Outputs.
  % num_var_in_comp
  base_mat_size  % [1, 0, 0] or its permutations, depends on which dimension is zero.
  mat_size_contribution  % 1x3 GVar
  num_block_contribution  % Always 1x3 of [0, 0, 0]. It should not be used.
  part_frac
  hash_penalty_term  % Always 1x3 cell array of 0. It should not be used.
  p_comp  % Always 1x3 cell array of 0.

  identifier
end
methods
  function obj = PartInfoZero()
    % Do nothing.
  end

  % Override Build: do not recursively build low-level parts.
  function obj = Build(obj, level, part_id, shape, identifier)
    global param_manager;
    obj.level = level;
    obj.power = 2 ^ (level - 1);
    obj.sum_col = 2 ^ level;
    obj.sum_half = obj.sum_col / 2;
    obj.part_id = part_id;
    obj.shape = shape;
    obj.identifier = identifier;

    % Rename the dimensions and store information.
    if shape(1) == 0
      obj.zero_dim = 1;
      obj.nonzero_dim_1 = 2;
      obj.nonzero_dim_2 = 3;
      obj.base_mat_size = [0, 0, 1];
    elseif shape(2) == 0
      obj.zero_dim = 2;
      obj.nonzero_dim_1 = 1;
      obj.nonzero_dim_2 = 3;
      obj.base_mat_size = [1, 0, 0];
    else
      obj.zero_dim = 3;
      obj.nonzero_dim_1 = 1;
      obj.nonzero_dim_2 = 2;
      obj.base_mat_size = [0, 1, 0];
    end
    % base_mat_size needs to multiply with part_frac and inner_prod_size.

    % Complete split distributions.
    obj.complete_split_id = cell(1, 3);
    for t = 1 : 3
      obj.complete_split_id{t} = ones(1, 3 ^ obj.power) * (-1);  % represents always zero
    end
    obj.complete_split_id{obj.zero_dim}(1) = -2;  % represents always one
    for id = 1 : 3 ^ obj.power
      arr = DecodeCSD(id, obj.power);
      if sum(arr) ~= obj.shape(obj.nonzero_dim_1)
        continue;
      end
      obj.complete_split_id{obj.nonzero_dim_1}(id) = param_manager.Register(1, 0, 1, [0, 1 / 100]);
      arr_opposite = ones(1, obj.power) * 2 - arr;
      id_opposite = EncodeCSD(arr_opposite);
      obj.complete_split_id{obj.nonzero_dim_2}(id_opposite) = obj.complete_split_id{obj.nonzero_dim_1}(id);
    end

    % Register linear constraint: sum of complete split distributions is 1.
    lincon_terms = {};
    for id = 1 : 3 ^ obj.power
      if obj.complete_split_id{obj.nonzero_dim_1}(id) == -1  % not possible to be -2
        continue;
      end
      lincon_terms{end + 1, 1} = obj.complete_split_id{obj.nonzero_dim_1}(id);
      lincon_terms{end, 2} = 1;
    end
    param_manager.AddLinearConstraintEq(lincon_terms, 1);
    
    % Placeholder variables
    obj.num_block_contribution = {[0, 0, 0], [0, 0, 0], [0, 0, 0]};
    obj.hash_penalty_term = {0, 0, 0};
    obj.p_comp = {0, 0, 0};
  end

  % Use heuristics to set initial parameters. We simply choose the complete split distribution that
  % maximizes the number of entries in the result inner product tensor.
  % CVX required for this function, and the user may select a default CVX solver before calling.
  function SetInitial(obj, ~)
    global q;
    cvx_begin quiet
      variables csd(1, 3 ^ obj.power);
      inner_prod_size = sum(entr(csd));
      for id = 1 : 3 ^ obj.power
        arr = DecodeCSD(id, obj.power);
        inner_prod_size = inner_prod_size + csd(id) * sum(arr == 1) * log(q);
      end
      maximize inner_prod_size;
      subject to
        sum(csd) == 1;
        csd >= 0;
        for id = 1 : 3 ^ obj.power
          arr = DecodeCSD(id, obj.power);
          if sum(arr) ~= obj.shape(obj.nonzero_dim_1)
            csd(id) == 0;
          end
        end
    cvx_end
    csd = max(csd, 0);
    csd = csd / sum(csd);
    global param_manager;
    for id = 1 : 3 ^ obj.power
      arr = DecodeCSD(id, obj.power);
      if sum(arr) ~= obj.shape(obj.nonzero_dim_1)
        continue;
      end
      param_manager.SetSingleParam(obj.complete_split_id{obj.nonzero_dim_1}(id), csd(id));
    end
  end

  % Override the evaluation process.
  function EvaluateInit(obj)
    global param_manager;
    obj.part_frac = GVar(param_manager.num_input, 0);
    obj.complete_split = cell(1, 3);
    for t = 1 : 3
      obj.complete_split{t} = GVar.Zeros(1, 3 ^ obj.power);
      for id = 1 : 3 ^ obj.power
        if obj.complete_split_id{t}(id) == -1
          obj.complete_split{t}(id) = 0;
        elseif obj.complete_split_id{t}(id) == -2
          obj.complete_split{t}(id) = 1;
        else
          obj.complete_split{t}(id) = param_manager.GetParam(obj.complete_split_id{t}(id));
        end
      end
    end
  end

  function EvaluatePre(obj)
    % Do nothing. We do not split, so do not propagate the fraction.
  end

  function EvaluatePost(obj)
    % Compute matrix size contribution. Should be multiplied by obj.part_frac.
    % Should equal to 2^entropy * q^(number of one), sum over CSD.
    global q;
    
    % First, without part_frac.
    inner_prod_size = obj.complete_split{obj.nonzero_dim_1}.Entropy();
    for id = 1 : 3 ^ obj.power
      arr = DecodeCSD(id, obj.power);
      inner_prod_size = inner_prod_size + obj.complete_split{obj.nonzero_dim_1}(id) * sum(arr == 1) * log(q);
    end

    % Combine with part_frac and base_mat_size.
    obj.mat_size_contribution = obj.part_frac * inner_prod_size * obj.base_mat_size;
  end
end
end

File: fast_mm_python/src/evaluation/PartInfoZero.py
"""
This is similar to PartInfo for level >= 3, but specialized for (i, j, k) containing zeros.
In this case, many variables are not used. We do not register for them.
"""

import numpy as np
from ..autograd import GVar
from ..complete_split import EncodeCSD, DecodeCSD


class PartInfoZero:
    """
    PartInfo for shapes containing zeros at level >= 3.
    Simplified compared to regular PartInfo.
    """
    
    def __init__(self):
        """Initialize empty PartInfoZero."""
        self.level = None
        self.power = None
        self.sum_col = None
        self.sum_half = None
        self.part_id = None
        self.shape = None
        self.zero_dim = None  # Which dimension is zero (1, 2, or 3)
        self.nonzero_dim_1 = None
        self.nonzero_dim_2 = None
        
        # Optimizable vars
        self.complete_split = None  # Output 1x3 list of CSD
        self.complete_split_id = None  # 1x3 list of parameter IDs
        
        # Outputs
        self.base_mat_size = None  # [1, 0, 0] or its permutations
        self.mat_size_contribution = None  # 1x3 GVar
        self.num_block_contribution = None  # Always [0, 0, 0]
        self.part_frac = None
        self.hash_penalty_term = None  # Always 0
        self.p_comp = None  # Always 0
        
        self.identifier = None
    
    def Build(self, level, part_id, shape, identifier, parts=None):
        """
        Build the PartInfoZero instance.
        
        Args:
            level: Level (>= 3)
            part_id: ID of this part
            shape: Shape tuple (i, j, k) with at least one zero
            identifier: Identifier tuple [parent_id, region]
            parts: Global parts dictionary (not used for zero shapes)
        """
        from ..autograd import get_param_manager
        param_manager = get_param_manager()
        
        self.level = level
        self.power = 2 ** (level - 1)
        self.sum_col = 2 ** level
        self.sum_half = self.sum_col // 2
        self.part_id = part_id
        self.shape = tuple(shape) if not isinstance(shape, tuple) else shape
        self.identifier = tuple(identifier) if not isinstance(identifier, tuple) else identifier
        
        # Determine which dimension is zero
        shape_array = np.array(self.shape)
        if shape_array[0] == 0:
            self.zero_dim = 0
            self.nonzero_dim_1 = 1
            self.nonzero_dim_2 = 2
            self.base_mat_size = np.array([0, 0, 1])
        elif shape_array[1] == 0:
            self.zero_dim = 1
            self.nonzero_dim_1 = 0
            self.nonzero_dim_2 = 2
            self.base_mat_size = np.array([1, 0, 0])
        else:  # shape_array[2] == 0
            self.zero_dim = 2
            self.nonzero_dim_1 = 0
            self.nonzero_dim_2 = 1
            self.base_mat_size = np.array([0, 1, 0])
        
        # Complete split distributions
        csd_size = 3 ** self.power
        self.complete_split_id = [None, None, None]
        
        for t in range(3):
            # -1 represents always zero, -2 represents always one
            self.complete_split_id[t] = np.full(csd_size, -1, dtype=int)
        
        # Zero dimension is always (0, 0, ..., 0) -> encoded as 1
        self.complete_split_id[self.zero_dim][0] = -2  # Always one
        
        # For nonzero dimensions, register parameters
        param_id_map = {}  # Map from CSD id to parameter id
        
        for csd_id in range(1, csd_size + 1):
            arr = DecodeCSD(csd_id, self.power)
            
            # Check if this CSD matches the nonzero dimension's constraint
            if np.sum(arr) != shape_array[self.nonzero_dim_1]:
                continue
            
            # Register parameter for this CSD
            param_id = param_manager.Register(1, lb=0, ub=1, initializer=(0, 0.01))
            self.complete_split_id[self.nonzero_dim_1][csd_id - 1] = param_id
            
            # The opposite dimension has the complementary CSD
            arr_opposite = 2 * np.ones(self.power, dtype=int) - arr
            csd_id_opposite = EncodeCSD(arr_opposite)
            self.complete_split_id[self.nonzero_dim_2][csd_id_opposite - 1] = param_id
        
        # Register linear constraint: sum of complete split distribution = 1
        lincon_entries = []
        for csd_id in range(csd_size):
            param_id = self.complete_split_id[self.nonzero_dim_1][csd_id]
            if param_id >= 0:  # Not -1 or -2
                lincon_entries.append((param_id, np.array([[1.0]])))
        
        if lincon_entries:
            param_manager.AddLinearConstraintEq(lincon_entries, np.array([1.0]))
        
        # Placeholder variables
        self.num_block_contribution = [np.array([0, 0, 0])] * 3
        self.hash_penalty_term = [0, 0, 0]
        self.p_comp = [0, 0, 0]
    
    def SetInitial(self, json=None):
        """
        Set initial parameters using heuristics.
        We simply choose the CSD that maximizes the number of entries in the result.
        
        Note: This is a simplified version without CVX optimization.
        """
        from ..autograd import get_param_manager
        from ..control import get_q
        
        param_manager = get_param_manager()
        q = get_q()
        
        # Simple heuristic: uniform distribution over valid CSDs
        csd_size = 3 ** self.power
        valid_csds = []
        valid_param_ids = []
        
        for csd_id in range(csd_size):
            param_id = self.complete_split_id[self.nonzero_dim_1][csd_id]
            if param_id >= 0:
                valid_csds.append(csd_id + 1)
                valid_param_ids.append(param_id)
        
        if valid_param_ids:
            # Uniform distribution
            uniform_prob = 1.0 / len(valid_param_ids)
            for param_id in valid_param_ids:
                param_manager.SetSingleParam(param_id, np.array([uniform_prob]))
    
    def EvaluateInit(self):
        """
        Initialize evaluation: load parameters into GVars.
        """
        from ..autograd import get_param_manager
        param_manager = get_param_manager()
        
        self.part_frac = GVar(param_manager.num_input, 0)
        self.complete_split = [None, None, None]
        
        csd_size = 3 ** self.power
        for t in range(3):
            self.complete_split[t] = GVar.Zeros(csd_size)
            
            for csd_id in range(csd_size):
                param_id = self.complete_split_id[t][csd_id]
                
                if param_id == -1:
                    # Always zero
                    self.complete_split[t].value[csd_id] = 0
                elif param_id == -2:
                    # Always one
                    self.complete_split[t].value[csd_id] = 1
                else:
                    # Optimizable parameter
                    param_gvar = param_manager.GetParam(param_id)
                    self.complete_split[t].value[csd_id] = param_gvar.value[0]
                    # Copy gradient information
                    self.complete_split[t].grad[:, csd_id] = param_gvar.grad[:, 0]
    
    def EvaluatePre(self):
        """
        Pre-evaluation step. We do not split, so do not propagate the fraction.
        """
        # Do nothing
        pass
    
    def EvaluatePost(self):
        """
        Post-evaluation: compute matrix size contribution.
        Should equal 2^entropy * q^(number of ones), summed over CSD.
        """
        from ..control import get_q
        q = get_q()
        
        # Compute inner product size
        inner_prod_size = self.complete_split[self.nonzero_dim_1].Entropy()
        
        csd_size = 3 ** self.power
        for csd_id in range(1, csd_size + 1):
            arr = DecodeCSD(csd_id, self.power)
            num_ones = np.sum(arr == 1)
            
            csd_prob = self.complete_split[self.nonzero_dim_1].value[csd_id - 1]
            if csd_prob > 0:
                inner_prod_size = inner_prod_size + csd_prob * num_ones * np.log(q)
        
        # Combine with part_frac and base_mat_size
        self.mat_size_contribution = self.part_frac * inner_prod_size * GVar(
            self.part_frac.num_input, self.base_mat_size
        )

